### ISSUE TITLE
cloud controller is unable to patch node with IP addresses from cloud provider

### ISSUE THREAD

**Issue by deshui123** [kubernetes/cloud-provider-vsphere#368]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
> /kind bug

**What happened**:
For IPv6(both internal and external are IPv6), cloud controller is unable to patch node with IP addresses from cloud provider on a multi-NIC VM.

I have added **internal-vm-network-name**, but this can not work for all nodes, even there is some logs look like 'Adding Internal IP by NetworkName: 2001:da8:8001:801::115'

```
[root@m6-g1-01 ~]# kubectl get node -o wide
NAME       STATUS   ROLES    AGE     VERSION   INTERNAL-IP              EXTERNAL-IP         OS-IMAGE                                      KERNEL-VERSION                CONTAINER-RUNTIME
m6-g1-01   Ready    <none>   5h27m   v1.18.6   2001:da8:8001:801::111   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g1-02   Ready    <none>   5h27m   v1.18.6   2001:da8:8001:801::112   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g1-03   Ready    <none>   5h27m   v1.18.6   2600:0:0:810::108        2600:0:0:810::108   Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g2-01   Ready    <none>   5h26m   v1.18.6   2600:0:0:810::109        2600:0:0:810::109   Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g2-02   Ready    <none>   5h26m   v1.18.6   2600:0:0:810::110        2600:0:0:810::110   Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g3-01   Ready    <none>   5h26m   v1.18.6   2001:da8:8001:801::116   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g3-02   Ready    <none>   5h26m   v1.18.6   2001:da8:8001:801::117   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
```

We can see that, **m6-g1-02** has the right internal IP, but for **m6-g2-02**, the internal IP is wrong, even there is some logs look like **Adding Internal IP by NetworkName: 2001:da8:8001:801::115**
Following is some logs about m6-g1-02 and m6-g2-02

```
I0803 08:02:49.916424       1 nodemanager.go:222] Adding Hostname: m6-g1-02
I0803 08:02:49.916497       1 nodemanager.go:282] Adding Internal IP by NetworkName: 2001:da8:8001:801::112
I0803 08:02:49.916536       1 nodemanager.go:306] Adding IP: 2600:0:0:810::107
I0803 08:02:49.916558       1 nodemanager.go:331] Found node 4214ab1b-2e28-c36a-6e86-0002790e78ac as vm=VirtualMachine:vm-755 in vc=2600:0:0:810::68 and datacenter=Datacenter1
I0803 08:02:49.916586       1 nodemanager.go:333] Hostname: m6-g1-02 UUID: 4214ab1b-2e28-c36a-6e86-0002790e78ac
I0803 08:02:49.916610       1 instances.go:163] instances.InstanceExistsByProviderID() EXISTS with 4214ab1b-2e28-c36a-6e86-0002790e78ac
I0803 08:02:49.916632       1 instances.go:78] instances.NodeAddressesByProviderID() CACHED with 4214ab1b-2e28-c36a-6e86-0002790e78ac
E0803 08:02:49.931003       1 node_controller.go:188] Error patching node with cloud ip addresses = [failed to patch status "{\"status\":{\"$setElementOrder/addresses\":[{\"type\":\"Hostname\"},{\"type\":\"InternalIP\"},{\"type\":\"ExternalIP\"},{\"type\":\"InternalIP\"}],\"addresses\":[{\"address\":\"2600:0:0:810::107\",\"type\":\"InternalIP\"},{\"address\":\"2001:da8:8001:801::112\",\"type\":\"InternalIP\"},{\"address\":\"2600:0:0:810::107\",\"type\":\"ExternalIP\"}]}}" for node "m6-g1-02": The order in patch list:
[map[address:2600:0:0:810::107 type:InternalIP] map[address:2001:da8:8001:801::112 type:InternalIP] map[address:2600:0:0:810::107 type:ExternalIP]]
 doesn't match $setElementOrder list:
[map[type:Hostname] map[type:InternalIP] map[type:ExternalIP] map[type:InternalIP]]
]
I0803 08:02:49.931167       1 search.go:68] WhichVCandDCByNodeID by UUID
I0803 08:02:49.931225       1 search.go:75] WhichVCandDCByNodeID nodeID: 421459f3-8568-f17d-ff43-8f7f79421063
I0803 08:02:49.943215       1 search.go:207] Found node 4214926f-8650-9ccd-3206-da585f010ad8 as vm=VirtualMachine:vm-758 in vc=2600:0:0:810::68 and datacenter=Datacenter1
I0803 08:02:49.943310       1 search.go:209] Hostname: m6-g2-02, UUID: 4214926f-8650-9ccd-3206-da585f010ad8
I0803 08:02:49.943409       1 nodemanager.go:123] Discovered VM using normal UUID format
I0803 08:02:49.951224       1 nodemanager.go:222] Adding Hostname: m6-g2-02
I0803 08:02:49.951332       1 nodemanager.go:306] Adding IP: 2600:0:0:810::110
I0803 08:02:49.951386       1 nodemanager.go:282] Adding Internal IP by NetworkName: 2001:da8:8001:801::115
I0803 08:02:49.951413       1 nodemanager.go:331] Found node 4214926f-8650-9ccd-3206-da585f010ad8 as vm=VirtualMachine:vm-758 in vc=2600:0:0:810::68 and datacenter=Datacenter1
I0803 08:02:49.951435       1 nodemanager.go:333] Hostname: m6-g2-02 UUID: 4214926f-8650-9ccd-3206-da585f010ad8
I0803 08:02:49.951486       1 search.go:68] WhichVCandDCByNodeID by UUID
I0803 08:02:49.951514       1 search.go:75] WhichVCandDCByNodeID nodeID: 4214ab1b-2e28-c36a-6e86-0002790e78ac
I0803 08:02:49.966868       1 search.go:207] Found node 421459f3-8568-f17d-ff43-8f7f79421063 as vm=VirtualMachine:vm-760 in vc=2600:0:0:810::68 and datacenter=Datacenter1
```

**What you expected to happen**:
1) GET the right internal IP for all nodes when run 'kubectl get node -o wide'
2) I have another question, why there are some logs look like this and what does this mean, how to resolve this?
```
E0803 08:02:49.931003       1 node_controller.go:188] Error patching node with cloud ip addresses = [failed to patch status "{\"status\":{\"$setElementOrder/addresses\":[{\"type\":\"Hostname\"},{\"type\":\"InternalIP\"},{\"type\":\"ExternalIP\"},{\"type\":\"InternalIP\"}],\"addresses\":[{\"address\":\"2600:0:0:810::107\",\"type\":\"InternalIP\"},{\"address\":\"2001:da8:8001:801::112\",\"type\":\"InternalIP\"},{\"address\":\"2600:0:0:810::107\",\"type\":\"ExternalIP\"}]}}" for node "m6-g1-02": The order in patch list:
[map[address:2600:0:0:810::107 type:InternalIP] map[address:2001:da8:8001:801::112 type:InternalIP] map[address:2600:0:0:810::107 type:ExternalIP]]
 doesn't match $setElementOrder list:
[map[type:Hostname] map[type:InternalIP] map[type:ExternalIP] map[type:InternalIP]]
```
**How to reproduce it (as minimally and precisely as possible)**:

Build a master node with 2 NICs which using pure IPv6 and attempt to deploy the cloud controller

**Anything else we need to know?**:

**Environment**:
- vsphere-cloud-controller-manager version:  Cloud Provider vSphere **v1.1.0**

```
[root@m6-g1-01 ~]# cat /etc/os-release
NAME="Red Hat Enterprise Linux Server"
VERSION="7.8 (Maipo)"
ID="rhel"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="7.8"
PRETTY_NAME="Red Hat Enterprise Linux Server 7.8 (Maipo)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:7.8:GA:server"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 7"
REDHAT_BUGZILLA_PRODUCT_VERSION=7.8
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="7.8"
```

**Comment by deshui123** [comment]
```
[root@m6-g1-01 ~]# cat /etc/kubernetes/cloud.conf
[Global]
insecure-flag = "1"
user = "****"
password = "*****"
port = 443
ip-family = "ipv6"
[VirtualCenter "2600:0:0:810::68"]
[Workspace]
datacenter = Datacenter1
server = 2600:0:0:810::68
folder = Cluster1
default-datastore = vsanDatastore
resourcepool-path = test_pool

[Disk]
scsicontrollertype = pvscsi

[Nodes]
internal-vm-network-name = Pri1-DPortGroup
```
According to this link: https://github.com/kubernetes/cloud-provider-vsphere/blob/870ee4bde1ca7056117b9b84e60ba32808260a3e/pkg/cloudprovider/vsphere/nodemanager.go,  I think only provide the internal-vm-network-name, CCM also can get the internal network
```
} else if internalVMNetworkName != "" && strings.EqualFold(internalVMNetworkName, v.Network) {
				for _, ip := range ips {
					klog.V(2).Infof("Adding Internal IP by NetworkName: %s", ip)
					v1helper.AddToNodeAddresses(&addrs,
						v1.NodeAddress{
							Type:    v1.NodeInternalIP,
							Address: ip,
						},
					)
					found = true
					break
				}
```

**Comment by dvonthenen** [comment]
HI @deshui123 The configuration you provided seems to be for the in-tree cloud provider. The documentation for using the external cloud provider can be found here:
https://cloud-provider-vsphere.sigs.k8s.io/tutorials/kubernetes-on-vsphere-with-kubeadm.html

**Comment by deshui123** [comment]
Hi @dvonthenen, 
Thanks for your response.
For the configuration, i think even if keep [Workspace] and [Disk] tag in cloud.conf, it will not cause any issues, since CPI will not use it. Please correct me if I  am wrong

As you mentioned, the CPI only supports a single IP address (v4 or v6) on a given NIC. In my lab, we only have one single IPv6 address for each NIC, and all nodes have the same network config, which means all nodes only have one single IPv6 address per each NIC.
Following is IP info for node m6-g2-02
```
[root@m6-g2-02 ~]# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:50:56:94:11:2a brd ff:ff:ff:ff:ff:ff
    inet6 2001:da8:8001:801::115/64 scope global noprefixroute
       valid_lft forever preferred_lft forever
    inet6 fe80::250:56ff:fe94:112a/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:50:56:94:a4:7d brd ff:ff:ff:ff:ff:ff
    inet6 2600:0:0:810::110/64 scope global noprefixroute
       valid_lft forever preferred_lft forever
    inet6 fe80::250:56ff:fe94:a47d/64 scope link noprefixroute
```
Also, I tried to remove the two tags from cloud.conf and still can not get all right internal_network_name.
```
[Global]
insecure-flag = "1"
user = "ncms@vsphere.clcm"
password = "Clcm2017$"
port = 443
ip-family = "ipv6"
[VirtualCenter "2600:0:0:810::68"]
datacenter = Datacenter1

[Nodes]
internal-vm-network-name = Pri1-DPortGroup
```
Some logs for m6-g2-02
```
I0804 02:35:11.241221       1 search.go:209] Hostname: m6-g2-02, UUID: 4214926f-8650-9ccd-3206-da585f010ad8
I0804 02:35:11.241251       1 nodemanager.go:123] Discovered VM using normal UUID format
I0804 02:35:11.258840       1 nodemanager.go:222] Adding Hostname: m6-g2-02
I0804 02:35:11.258972       1 nodemanager.go:306] Adding IP: 2600:0:0:810::110
I0804 02:35:11.259026       1 nodemanager.go:282] Adding Internal IP by NetworkName: 2001:da8:8001:801::115
I0804 02:35:11.259045       1 nodemanager.go:331] Found node 4214926f-8650-9ccd-3206-da585f010ad8 as vm=VirtualMachine:vm-758 in vc=2600:0:0:810::68 and datacenter=Datacenter1
I0804 02:35:11.259079       1 nodemanager.go:333] Hostname: m6-g2-02 UUID: 4214926f-8650-9ccd-3206-da585f010ad8
I0804 02:35:11.259130       1 search.go:68] WhichVCandDCByNodeID by UUID
I0804 02:35:11.259145       1 search.go:75] WhichVCandDCByNodeID nodeID: 4214ab1b-2e28-c36a-6e86-0002790e78ac
I0804 02:35:11.306175       1 search.go:207] Found node 4214ab1b-2e28-c36a-6e86-0002790e78ac as vm=VirtualMachine:vm-755 in vc=2600:0:0:810::68 and datacenter=Datacenter1
```
output of kubectl
```
[root@m6-g1-03 ~]# kubectl get node -o wide
NAME       STATUS   ROLES    AGE   VERSION   INTERNAL-IP              EXTERNAL-IP         OS-IMAGE                                      KERNEL-VERSION                CONTAINER-RUNTIME
m6-g1-01   Ready    <none>   23h   v1.18.6   2001:da8:8001:801::111   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g1-02   Ready    <none>   23h   v1.18.6   2001:da8:8001:801::112   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g1-03   Ready    <none>   23h   v1.18.6   2600:0:0:810::108        2600:0:0:810::108   Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g2-01   Ready    <none>   23h   v1.18.6   2600:0:0:810::109        2600:0:0:810::109   Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g2-02   Ready    <none>   23h   v1.18.6   2600:0:0:810::110        2600:0:0:810::110   Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g3-01   Ready    <none>   23h   v1.18.6   2001:da8:8001:801::116   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g3-02   Ready    <none>   23h   v1.18.6   2001:da8:8001:801::117   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
```
I tried to add internal-network-subnet-cidr to cloud.conf, still failed. I think if only one IP for one NIC, the internal-network-subnet-cidr is not needed.
```
[Global]
insecure-flag = "1"
user = "ncms@vsphere.clcm"
password = "Clcm2017$"
port = 443
ip-family = "ipv6"
[VirtualCenter "2600:0:0:810::68"]
datacenter = Datacenter1

[Nodes]
internal-vm-network-name = Pri1-DPortGroup
internal-network-subnet-cidr = 2001:da8:8001:801::/64
```
Some logs for m6-g2-02, we can see **Adding Internal IP by NetworkName: 2001:da8:8001:801::115** and **Successfully initialized node m6-g2-02 with cloud provider**
```
I0804 02:49:45.380523       1 search.go:209] Hostname: m6-g2-02, UUID: 4214926f-8650-9ccd-3206-da585f010ad8
I0804 02:49:45.380564       1 nodemanager.go:123] Discovered VM using normal UUID format
I0804 02:49:45.388759       1 nodemanager.go:222] Adding Hostname: m6-g2-02
I0804 02:49:45.388850       1 nodemanager.go:306] Adding IP: 2600:0:0:810::110
I0804 02:49:45.388885       1 nodemanager.go:282] Adding Internal IP by NetworkName: 2001:da8:8001:801::115
I0804 02:49:45.388948       1 nodemanager.go:331] Found node 4214926f-8650-9ccd-3206-da585f010ad8 as vm=VirtualMachine:vm-758 in vc=2600:0:0:810::68 and datacenter=Datacenter1
I0804 02:49:45.388973       1 nodemanager.go:333] Hostname: m6-g2-02 UUID: 4214926f-8650-9ccd-3206-da585f010ad8
I0804 02:49:45.389015       1 search.go:68] WhichVCandDCByNodeID by UUID
I0804 02:49:45.389040       1 search.go:75] WhichVCandDCByNodeID nodeID: 4214ab1b-2e28-c36a-6e86-0002790e78ac
I0804 02:49:45.421600       1 search.go:207] Found node 4214ab1b-2e28-c36a-6e86-0002790e78ac as vm=VirtualMachine:vm-755 in vc=2600:0:0:810::68 and datacenter=Datacenter1

I0804 02:49:46.019241       1 instances.go:158] instances.InstanceExistsByProviderID() CACHED with 4214926f-8650-9ccd-3206-da585f010ad8
I0804 02:49:46.019291       1 instances.go:78] instances.NodeAddressesByProviderID() CACHED with 4214926f-8650-9ccd-3206-da585f010ad8
I0804 02:49:46.032451       1 node_controller.go:315] Successfully initialized node m6-g2-02 with cloud provider
```
logs of kubectl
```
[root@m6-g1-03 ~]# kubectl get node -o wide
NAME       STATUS   ROLES    AGE   VERSION   INTERNAL-IP              EXTERNAL-IP         OS-IMAGE                                      KERNEL-VERSION                CONTAINER-RUNTIME
m6-g1-01   Ready    <none>   24h   v1.18.6   2001:da8:8001:801::111   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g1-02   Ready    <none>   24h   v1.18.6   2001:da8:8001:801::112   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g1-03   Ready    <none>   24h   v1.18.6   2600:0:0:810::108        2600:0:0:810::108   Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g2-01   Ready    <none>   24h   v1.18.6   2600:0:0:810::109        2600:0:0:810::109   Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g2-02   Ready    <none>   24h   v1.18.6   2600:0:0:810::110        2600:0:0:810::110   Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g3-01   Ready    <none>   24h   v1.18.6   2001:da8:8001:801::116   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
m6-g3-02   Ready    <none>   24h   v1.18.6   2001:da8:8001:801::117   <none>              Red Hat Enterprise Linux Server 7.8 (Maipo)   4.18.0-147.8.1.el8_1.x86_64   docker://1.13.1
```

**Comment by deshui123** [comment]
I check some code in cloud-provider-openstack, hope these fix can provide some idea for you
https://github.com/kubernetes/cloud-provider-openstack/pull/1041
https://github.com/kubernetes/cloud-provider-openstack/commit/c544421eaeaf8124786f2ef54cfaa6f794b1fe87
https://github.com/kubernetes/cloud-provider-openstack/blame/6d0ea905cabb5e8a974c4bdfb49d1b354907822a/pkg/cloudprovider/providers/openstack/openstack.go
we can see  the comments **removing already added address to avoid listing it as both ExternalIP and InternalIP**
```
} else if util.Contains(networkingOpts.PublicNetworkName, network) {
				addressType = v1.NodeExternalIP
				// removing already added address to avoid listing it as both ExternalIP and InternalIP
				// may happen due to listing "private" network as "public" in CCM's config
				RemoveFromNodeAddresses(&addrs,
					v1.NodeAddress{
						Address: props.Addr,
					},
				)
			} else {
				if len(networkingOpts.InternalNetworkName) == 0 || util.Contains(networkingOpts.InternalNetworkName, network) {
					addressType = v1.NodeInternalIP
				} else {
					klog.V(5).Infof("Node '%s' address '%s' ignored due to 'internal-network-name' option", srv.Name, props.Addr)
					RemoveFromNodeAddresses(&addrs,
						v1.NodeAddress{
							Address: props.Addr,
						},
					)
```

**Comment by deshui123** [comment]
Also hit this issue for **IPV4**
```
[root@tesla-auto-2006-v67-os7-v22-control-01 ~]# kubectl get nodes -o wide
NAME                                     STATUS   ROLES    AGE    VERSION   INTERNAL-IP       EXTERNAL-IP       OS-IMAGE                KERNEL-VERSION                CONTAINER-RUNTIME
tesla-auto-2006-v67-os7-v22-control-01   Ready    <none>   6h4m   v1.18.6   135.111.166.138   135.111.166.138   CentOS Linux 7 (Core)   4.18.0-147.5.1.el8_1.x86_64   docker://19.3.12
tesla-auto-2006-v67-os7-v22-control-02   Ready    <none>   6h4m   v1.18.6   135.111.166.139   135.111.166.139   CentOS Linux 7 (Core)   4.18.0-147.5.1.el8_1.x86_64   docker://19.3.12
tesla-auto-2006-v67-os7-v22-control-03   Ready    <none>   6h4m   v1.18.6   172.26.1.23       <none>            CentOS Linux 7 (Core)   4.18.0-147.5.1.el8_1.x86_64   docker://19.3.12
tesla-auto-2006-v67-os7-v22-edge-01      Ready    <none>   6h3m   v1.18.6   172.26.1.26       <none>            CentOS Linux 7 (Core)   4.18.0-147.5.1.el8_1.x86_64   docker://19.3.12
tesla-auto-2006-v67-os7-v22-edge-02      Ready    <none>   6h3m   v1.18.6   135.111.166.144   135.111.166.144   CentOS Linux 7 (Core)   4.18.0-147.5.1.el8_1.x86_64   docker://19.3.12
tesla-auto-2006-v67-os7-v22-worker-01    Ready    <none>   6h3m   v1.18.6   172.26.1.24       <none>            CentOS Linux 7 (Core)   4.18.0-147.5.1.el8_1.x86_64   docker://19.3.12
tesla-auto-2006-v67-os7-v22-worker-02    Ready    <none>   6h3m   v1.18.6   135.111.166.142   135.111.166.142   CentOS Linux 7 (Core)   4.18.0-147.5.1.el8_1.x86_64   docker://19.3.12
```
```
[cloud-user@tesla-auto-1912-os7-v2-control-03 ~]$ ip a1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:50:56:a9:27:e4 brd ff:ff:ff:ff:ff:ff
    inet 135.1.136.12/27 brd 135.1.136.31 scope global noprefixroute eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::250:56ff:fea9:27e4/64 scope link
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:50:56:a9:54:65 brd ff:ff:ff:ff:ff:ff
    inet 172.26.1.23/24 brd 172.26.1.255 scope global noprefixroute eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::250:56ff:fea9:5465/64 scope link
       valid_lft forever preferred_lft forever
```

**Comment by deshui123** [comment]
any update?

**Comment by dvonthenen** [comment]
I'm looking at this right now and it looks like there might be a bug. Will post back shortly.

**Comment by dvonthenen** [comment]
@deshui123 If you want to give this a try, the latest image is located here:
dvonthenen/vsphere-cloud-controller-manager:latest

It's based on this code:
https://github.com/dvonthenen/cloud-provider-vsphere/tree/bug368/addyreporting

**Comment by deshui123** [comment]
Hi @dvonthenen,
Thanks for your response.
Usually, we use the release from https://github.com/kubernetes/cloud-provider-vsphere. Hope we can get the latest release which covers this fix ASAP.

**Comment by dvonthenen** [comment]
This is untested and I currently need to find the time to try this out (not sure when that might be)... which is where I would hope you might come in since you are having trouble. If not, I will give this a try when I get my testbed up and running again.

**Comment by dvonthenen** [comment]
@deshui123 PR https://github.com/kubernetes/cloud-provider-vsphere/pull/378 should have addressed in the 1.2.1 release. Please give this a try. If it doesn't meet you needs, please let us know.

/close
